woocommerce/facebook-for-woocommerce
# add query monitor profiling to three specific product-sync code paths

This PR adds ~~[Query Monitor profiling](https://github.com/johnbillion/query-monitor#profiling)~~ profiling to three aspects of product sync. (Now switched to use the profiler from partner PR #1854.)

- Get product ID list for background batch sync / feed `facebook_for_woocommerce__create_or_update_all_products`
- Process background batch sync `facebook_for_woocommerce__background_product_sync__process_job`
- Feed file generation `facebook_for_woocommerce__generate_feed`

~~We may not end up merging this PR. It's intended for helping us measure & optimise product sync. I'm opening this PR for visibility only for now. If this proves useful (and is harmless in production), we could consider merging in future (or something similar).~~ Update: keen to get this shipshape and merged. **When reviewing please consider implications of adding this to the plugin.**

To test:

- [Ensure profiler logging is enabled](https://github.com/woocommerce/facebook-for-woocommerce/tree/master/includes/Debug) `define( 'FACEBOOK_FOR_WOOCOMMERCE_PROFILING_LOG_ENABLED', true );`. I also had debug mode active: `Marketing > Facebook > Connection tab > Log plugin events for debugging`.
- Go to `Marketing > Facebook > Product sync` tab. 
- Click `Sync products` to trigger a sync.
- View logs, e.g. `WooCommerce > Status > Logs`. There should be a `facebook_for_woocommerce_profiling_*` log.

Should see stuff like this in the log file:

```
2021-04-08T02:54:05+00:00 DEBUG create_or_update_all_products - Memory: 6,334.59 KB, time: 0.02s.
2021-04-08T02:54:32+00:00 DEBUG background_product_sync__process_job - Memory: 131,754.52 KB, time: 26.41s.
2021-04-08T02:54:54+00:00 DEBUG background_product_sync__process_job - Memory: 16,114.16 KB, time: 21.52s.
```
